### PR TITLE
Add autogeneration of block list reference documentation

### LINKED ISSUE #45373: Prepare documentation of available blocks
**Background:**

There's no documentation that would list all of the blocks.

**Goal:**

1. Prepare such document (name, label, description)
2. Worth aiming to automate that process and make the list dynamic so it's always up to date (need to include some sort of information about deprecated blocks, experimental ones etc.)

### PR DESCRIPTION
**Pull request by roykho**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/45373

This PR adds a script to autogenerate the block reference to the documentation.

I searched around before I started as I didn't want to "reinvent the wheel" and I found https://github.com/WordPress/gutenberg/pull/36183. So I adopted this script for our use.

Example document output:
https://github.com/woocommerce/woocommerce/blob/blocks/reference-documentation/plugins/woocommerce-blocks/docs/block-references/block-references.md

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**To be tested by developers only.**

1. Go to project root.
2. Run `pnpm --filter="@woocommerce/block-library" build:docs:block-references`.
3. You should see the generated doc in `./plugins/woocommerce-blocks/docs/block-references/block-references.md`

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

Adds autogeneration of our block list references.
</details>

### CODE REVIEW

**Comment by roykho** [comment]
You will notice failing markdown linting `Multiple headings with the same content`. I am not sure what is the best way to handle that. Should block names be re-used? If that's the case, we need a way to ignore that rule. If not, then we need to update the names for those blocks.

**Comment by roykho** [comment]
@kmanijak - I've made some adjustments per our discussion yesterday.

- Added the name to the title to prevent duplicate lint warnings
- Added `ancestor` section
- If a key value is not set, don't display `undefined`

File: plugins/woocommerce-blocks/bin/gen-block-list-doc.js [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	return `
+## ${ blockjson.title } - ${ blockjson.name }
+
+${ blockjson.description || '' }
+
+-	**Name:** ${ blockjson.name }
+-	**Category:** ${ blockjson.category || '' }
+-   **Ancestor:** ${ blockjson.ancestor || '' }
```
Comment (kmanijak) [inline plugins/woocommerce-blocks/bin/gen-block-list-doc.js:164]: I just realised some blocks specify `parent` metadata which makes them available only when directly nested within specific block. Worth adding that field as well.

```suggestion
-   **Ancestor:** ${ blockjson.ancestor || '' }
-   **Parent:** ${ blockjson.parent || '' }
```

File: plugins/woocommerce-blocks/docs/block-references/block-references.md [hunk 2]
```diff
@@ -0,0 +1,1187 @@
+# Woo Blocks Reference
+
+This page lists the Woo blocks included in the package.
```
Comment (kmanijak) [inline plugins/woocommerce-blocks/docs/block-references/block-references.md]: I think it's worth mentioning the list may be incomplete due to some (usually older) blocks being registered without `block.json` and hence omited by script.

**Review by kmanijak — APPROVED** [review]
Thanks for handling that! 🚀 I left some minor comments but want to raise some other points.

#### Incomplete list
While I think it's ok we have incomplete list of blocks (as long as we mention this in doc), what about creating a followup issue to "migrate" all the remaining blocks definition to `block.json` so they can be recognized by script?

#### Experimental blocks
I don't have a good suggestion how to handle it but now doc includes the experimental blocks as well. And these blocks are not available to the end users so that may be misleading. Would be great if we could either skip them or mark them as experimental.

One option is we could utilise `version` metadata in block definition. Right now we just put `1.0.0` and keep it that way no matter what happens with a block. We could follow a pseudo-semver and do this:
- `0.1.x` - block is experimental
- `1.0.0-beta-x` - block is public in Beta
- `x.0.0` - block is public and stable

In that case we could easily recognise the block state and skip experimental ones.

Another idea is we simply add `(Experimental)` to blocks title. Right now Product Filters are experimental but have `(Beta)` in title which is misleading even for the team members.

Anyway, that requires RFC and agreement among all team members. I'll try to write it before AFK but can't promise I'll make it, otherwise feel free to do it while I'm away. Having said that, I'm pre-approving but please address two minor comments.

**Review by kmanijak — APPROVED** [review]
Thanks for handling that! 🚀 I left some minor comments but want to raise some other points.

#### Incomplete list
While I think it's ok we have incomplete list of blocks (as long as we mention this in doc), what about creating a followup issue to "migrate" all the remaining blocks definition to `block.json` so they can be recognized by script?

#### Experimental blocks
I don't have a good suggestion how to handle it but now doc includes the experimental blocks as well. And these blocks are not available to the end users so that may be misleading. Would be great if we could either skip them or mark them as experimental.

One option is we could utilise `version` metadata in block definition. Right now we just put `1.0.0` and keep it that way no matter what happens with a block. We could follow a pseudo-semver and do this:
- `0.1.x` - block is experimental
- `1.0.0-beta-x` - block is public in Beta
- `x.0.0` - block is public and stable

In that case we could easily recognise the block state and skip experimental ones.

Another idea is we simply add `(Experimental)` to blocks title. Right now Product Filters are experimental but have `(Beta)` in title which is misleading even for the team members.

Anyway, that requires RFC and agreement among all team members. I'll try to write it before AFK but can't promise I'll make it, otherwise feel free to do it while I'm away. Having said that, I'm pre-approving but please address two minor comments.

**Comment by roykho** [comment]
Created issue to migrate the rest of the blocks to use `block.json` https://github.com/woocommerce/woocommerce/issues/46917

**Pull request merged**